lucille-core/src/main/java/com/kmwllc/lucille/util/SolrUtils.java

```diff
@@ -117,21 +119,58 @@ public static List<String> getSolrUrls(Config config) {
 
   /**
    * This method will convert a Tuple produced by a Solr Streaming Expression into a Lucille Document. The Tuple must
-   * have a value in its id field in order to be converted.
+   * have a value in its id field in order to be converted. The types of values are maintained in the conversion. Empty list 
+   * are not included in the created document
    *
    * @param tuple a Tuple from Solr
-   * @return a Document
+   * @throws DocumentException if {@param tuple} does not have a key called Document.ID_FIELD
+   * @return a Document whose id is the stringified version of the contents of Document.ID_FIELD in the tuple.
    */
   public static Document toDocument(Tuple tuple) throws DocumentException {
     Document d;
-    if (tuple.getString(Document.ID_FIELD) != null) {
+
+    if (tuple.getFields().containsKey(Document.ID_FIELD)) {
       d = Document.create(tuple.getString(Document.ID_FIELD));
     } else {
       throw new DocumentException("Unable to create Document from Tuple. No id field present.");
     }
 
     for (Entry<String, Object> e : tuple.getFields().entrySet()) {
-      d.setOrAdd((String) e.getKey(), (String) e.getValue());
+      Object val = e.getValue();
+      String key = e.getKey();
+
+      if (key.equals(Document.ID_FIELD)) {
+        continue;
+      }
+
+      if (val instanceof String) {
```

Comment:
This code could be expanded to cover the remaining types that Document supports and moved into a method on the Document interface itself:   `default void setOrAdd(String key, Object val) throws DocumentException`

That would be a big enough change to warrant moving this to a separate PR with its own ticket

I am trying to remember if there's a reason we didn't do something like this earlier. In any case, it seems better to have this type inference logic encapsulated in Document rather than living just in SolrUtils. LMK what you think.

Reply:
I was thinking of an expanded version of something like this on Document:

```
  default void setOrAdd(String key, Object val) throws DocumentException {
    if (val instanceof String) {
      setOrAdd(key, (String) val);
    } else if (val instanceof Long) {
      setOrAdd(key, (Long) val);
    } else if (val instanceof Double) {
      setOrAdd(key, (Double) val);
    } else if (val instanceof Boolean) {
      setOrAdd(key, (Boolean) val);
    } else if (val instanceof List<?>) {
      List<?> list = (List<?>) val;
      if (!list.isEmpty()) {
        if (list.get(0) instanceof String) {
          update(key, UpdateMode.APPEND, list.toArray(new String[0]));
        } else if (list.get(0) instanceof Long) {
          update(key, UpdateMode.APPEND, list.toArray(new Long[0]));
        } else if (list.get(0) instanceof Double) {
          update(key, UpdateMode.APPEND, list.toArray(new Double[0]));
        } else if (list.get(0) instanceof Boolean) {
          update(key, UpdateMode.APPEND, list.toArray(new Boolean[0]));
        } else {
          throw new DocumentException(String.format("Class %s is not supported in Document", val.getClass()));
        }
      }
    }
  }
```

Reply:
I think the instanceof checks are, plain and simple, what we need here. I'm not thinking of a more elegant approach... The good news is that all this logic would be encapsulated in one method on Document. I was trying to recall whether we had type checking code like this elsewhere in the codebase, because I'm sure this need has arisen before, but I don't think we do. I would say go ahead with adding this as a default method on Document, and we can refactor that method later if we find a better approach. This work will also help us with an existing limitation of the DatabaseConnector -- currently it treats everything as a string. See https://github.com/kmwtechnology/lucille/blob/7418a2423a91b6eef8bdefc8cbd6783ea1d42ae0/lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java#L188

There's a related ticket for DatabaseConnector referencing a returnTypes array but that appears to be stale:
https://kmwllc.atlassian.net/browse/LC-220

Reply:
i refactored the SolrUtils.toDocument method to use the generic Document setters which have been added. The method now supports adding all types supported by Document as well as Lists and Arrays of those types.